File: Transport/Stateflows.Transport.AspNetCore.SignalR.Client/Behavior.cs
using Microsoft.AspNetCore.SignalR.Client;
using Stateflows.Common;
using Stateflows.Common.Utilities;
using Stateflows.Common.Exceptions;

namespace Stateflows.Transport.SignalR.Client
{
    internal class Behavior : IBehavior
    {
        private Task<HubConnection> Hub { get; }

        private async Task<HubConnection> GetHub()
        {
            var result = await Hub;

            if (result.State != HubConnectionState.Connected)
            {
                throw new TransportException("Transport failure: SignalR connection unavailable");
            }

            return result;
        }

        public BehaviorId Id { get; }

        public Behavior(Task<HubConnection> hub, BehaviorId id)
        {
            Hub = hub;
            Id = id;
        }

        public async Task<SendResult> SendAsync<TEvent>(TEvent @event, IEnumerable<EventHeader> headers = null)
        {
            var eventHolder = @event.ToTypedEventHolder();
            eventHolder.Headers.AddRange(headers);

            ResponseHolder.SetResponses(new Dictionary<object, EventHolder>());

            var hub = await GetHub();

            var resultString = await hub.InvokeAsync<string>("Send", Id, StateflowsJsonConverter.SerializePolymorphicObject(eventHolder, true));

            var result = StateflowsJsonConverter.DeserializeObject<RequestResult>(resultString);

            return new SendResult(eventHolder, result.Status, Array.Empty<EventHolder>(), result.Validation);
        }

        public async Task<RequestResult<TResponse>> RequestAsync<TResponse>(IRequest<TResponse> request, IEnumerable<EventHeader> headers = null)
        {
            var requestHolder = request.ToTypedEventHolder();
            requestHolder.Headers.AddRange(headers);

            ResponseHolder.SetResponses(new Dictionary<object, EventHolder>());

            var hub = await GetHub();

            var resultString = await hub.InvokeAsync<string>("Request", Id, StateflowsJsonConverter.SerializePolymorphicObject(request, true));

            var result = StateflowsJsonConverter.DeserializeObject<RequestResult>(resultString);

            return new RequestResult<TResponse>(requestHolder, result.Status, result.Validation);
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        { }

        public Task<IWatcher> WatchAsync<TNotification>(Action<TNotification> handler)
        {
            throw new NotImplementedException();
        }

        public Task<IWatcher> UnwatchAsync<TNotification>()
        {
            throw new NotImplementedException();
        }

        ~Behavior()
            => Dispose(false);
    }
}

File: Transport/Stateflows.Transport.TypeScript/@stateflows/client/src/behaviors/behavior.ts
import { HubConnection } from "@microsoft/signalr";
import { Request } from "../events/request";
import { IBehavior } from "../interfaces/behavior";

export class Behavior implements IBehavior {
    private hubPromise: Promise<HubConnection>;
    
    constructor(hubPromiseOrBehavior: Promise<HubConnection> | Behavior) {
        this.hubPromise = hubPromiseOrBehavior instanceof Behavior
            ? hubPromiseOrBehavior.hubPromise
            : this.hubPromise = hubPromiseOrBehavior;
    }

    send(event: Event): Promise<boolean> {
        throw new Error("Method not implemented.");
    }

    request<TResponse>(request: Request<TResponse>): Promise<TResponse> {
        throw new Error("Method not implemented.");
    }

    initialize(): Promise<boolean> {
        throw new Error("Method not implemented.");
    }

    isInitialized(): Promise<boolean> {
        throw new Error("Method not implemented.");
    }
}